File: local-movie-app/src/main/java/com/github/rahmnathan/localmovies/app/adapter/list/ListAdapterUtils.java
package com.github.rahmnathan.localmovies.app.adapter.list;

import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.util.Base64;
import android.view.Gravity;
import android.widget.ImageView;
import android.widget.TextView;

import com.github.rahmnathan.localmovies.app.data.Media;
import com.github.rahmnathan.localmovies.app.data.MovieOrder;

import java.util.Comparator;
import java.util.List;
import java.util.Objects;

import rahmnathan.localmovies.R;

public class ListAdapterUtils {

    static void sort(List<Media> media, MovieOrder order) {
        switch (order) {
            case DATE_ADDED:
                media.sort((movie1, movie2) -> Objects.requireNonNull(movie2.getCreated()).compareTo(movie1.getCreated()));
                break;
            case MOST_VIEWS:
                media.sort((movie1, movie2) -> Integer.compare(movie2.getViews(), movie1.getViews()));
                break;
            case RATING:
                media.sort((movie1, movie2) -> movie2.getImdbRating().compareTo(movie1.getImdbRating()));
                break;
            case RELEASE_YEAR:
                media.sort((movie1, movie2) -> movie2.getReleaseYear().compareTo(movie1.getReleaseYear()));
                break;
            case TITLE:
                media.sort(Comparator.comparing(Media::getTitle));
                break;
        }
    }

    public static void mapTitleToView(String title, TextView titleView, int fontSize){
        titleView.setText(title);
        titleView.setTextSize(fontSize);
        titleView.setGravity(Gravity.CENTER);
        titleView.setTextColor(Color.WHITE);
    }

    static void mapRatingsToView(String imdbRating, String metaRating, TextView ratings){
        ratings.setGravity(Gravity.CENTER);
        ratings.setTextColor(Color.WHITE);
        ratings.setTextSize(12);
        ratings.setText(String.format("IMDB: %s Meta: %s", imdbRating, metaRating));
    }

    public static void mapImageToView(String base64Image, ImageView imageView){
        if (base64Image != null && !base64Image.equals("") && !base64Image.equals("null")) {
            byte[] image = Base64.decode(base64Image, Base64.DEFAULT);
            imageView.setImageBitmap(BitmapFactory.decodeByteArray(image, 0, image.length));
        } else {
            imageView.setImageResource(R.mipmap.no_poster);
        }
    }

    public static void mapYearToView(String releaseYear, TextView year, int fontSize){
        year.setText(String.format("Release Year: %s", releaseYear));
        year.setTextColor(Color.WHITE);
        year.setGravity(Gravity.CENTER);
        year.setTextSize(fontSize);
    }
}

File: local-movie-app/src/main/java/com/github/rahmnathan/localmovies/app/activity/main/view/ListAdapterUtils.kt
package com.github.rahmnathan.localmovies.app.activity.main.view

import android.graphics.BitmapFactory
import android.graphics.Color
import android.util.Base64
import android.view.Gravity
import android.view.MenuItem
import android.widget.GridView
import android.widget.ImageView
import android.widget.TextView
import com.github.rahmnathan.localmovies.app.media.data.Media
import com.github.rahmnathan.localmovies.app.media.data.MediaGenre
import com.github.rahmnathan.localmovies.app.media.data.MediaOrder
import rahmnathan.localmovies.R

object ListAdapterUtils {

    @JvmStatic
    fun sort(mediaList: MutableList<Media>, order: MediaOrder?) {
        when (order) {
            MediaOrder.DATE_ADDED -> mediaList.sortByDescending { media -> media.created }
            MediaOrder.RATING -> mediaList.sortByDescending { media -> media.imdbRating }
            MediaOrder.RELEASE_YEAR -> mediaList.sortByDescending { media -> media.releaseYear }
            MediaOrder.TITLE -> mediaList.sortBy { media -> media.title }
            MediaOrder.NUMBER -> mediaList.sortBy { media -> Integer.parseInt(media.number!!) }
            else -> {}
        }
    }

    @JvmStatic
    fun mapTitleToView(title: String?, titleView: TextView, fontSize: Int) {
        titleView.text = title
        titleView.textSize = fontSize.toFloat()
        titleView.gravity = Gravity.CENTER
        titleView.setTextColor(Color.WHITE)
    }

    @JvmStatic
    fun mapRatingsToView(imdbRating: String?, metaRating: String?, ratings: TextView) {
        ratings.gravity = Gravity.CENTER
        ratings.setTextColor(Color.WHITE)
        ratings.textSize = 12f
        ratings.text = String.format("IMDB: %s Meta: %s", imdbRating, metaRating)
    }

    @JvmStatic
    fun mapImageToView(base64Image: String?, imageView: ImageView) {
        if (base64Image != null && base64Image != "" && base64Image != "null") {
            val image = Base64.decode(base64Image, Base64.DEFAULT)
            imageView.setImageBitmap(BitmapFactory.decodeByteArray(image, 0, image.size))
        } else {
            imageView.setImageResource(R.mipmap.no_poster)
        }
    }

    @JvmStatic
    fun mapYearToView(releaseYear: String?, year: TextView, fontSize: Int) {
        year.text = String.format("Release Year: %s", releaseYear)
        year.setTextColor(Color.WHITE)
        year.gravity = Gravity.CENTER
        year.textSize = fontSize.toFloat()
    }

    @JvmStatic
    fun sortVideoList(item: MenuItem, listAdapter: MediaListAdapter, gridView: GridView) {
        when (item.itemId) {
            R.id.order_date_added -> sort(MediaOrder.DATE_ADDED, listAdapter, gridView)
            R.id.order_year -> sort(MediaOrder.RELEASE_YEAR, listAdapter, gridView)
            R.id.order_rating -> sort(MediaOrder.RATING, listAdapter, gridView)
            R.id.order_title -> sort(MediaOrder.TITLE, listAdapter, gridView)
            R.id.genre_comedy -> filterGenre(MediaGenre.COMEDY, listAdapter, gridView)
            R.id.action_action -> filterGenre(MediaGenre.ACTION, listAdapter, gridView)
            R.id.genre_sciFi -> filterGenre(MediaGenre.SCIFI, listAdapter, gridView)
            R.id.genre_horror -> filterGenre(MediaGenre.HORROR, listAdapter, gridView)
            R.id.genre_thriller -> filterGenre(MediaGenre.THRILLER, listAdapter, gridView)
            R.id.genre_fantasy -> filterGenre(MediaGenre.FANTASY, listAdapter, gridView)
        }
    }

    private fun sort(order: MediaOrder, listAdapter: MediaListAdapter, gridView: GridView) {
        listAdapter.sort(order)
        gridView.smoothScrollToPosition(0)
    }

    private fun filterGenre(genre: MediaGenre, listAdapter: MediaListAdapter, gridView: GridView) {
        listAdapter.filterGenre(genre)
        gridView.smoothScrollToPosition(0)
    }
}